docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+```python
+data = ["large_document_1", "large_document_2"]
+pipeline = Embed(Chunk())
+collection = client.create_collection("my_collection", pipeline=pipeline)
+collection.add(data=data, ids=[0, 1])
+
+# Input to pipeline would be {ids: [0, 1], documents: ["large_document_1", "large_document_2"], metadata: None, embeddings: None}
+# Output of Chunk could be {ids: [A, B, C, D], documents: ["chunk_1", "chunk_2", "chunk_3", "chunk_4"], metadata: [{"source_id": 0}, {"source_id": 0}, {"source_id": 1}, {"source_id": 1}], embeddings: None}
+# Output of Embed, the final stage could be {ids: [0, 1], documents: ["chunk_1", "chunk_2", "chunk_3", "chunk_4"], metadata: [{"source_id": 0}, {"source_id": 0}, {"source_id": 1}, {"source_id": 1}], embeddings: [embedding_1, embedding_2, embedding_3, embedding_4]}
+```
+
+
+#### Pipeline overrides
+While defined at the collection level, pipelines can also be overridden at the time of an insert or query call. This allows for flexibility in the pipeline that is used for a specific insert or query call. For example, if you wanted to chunk the data initially, but later insert a custom chunk, you could do something like this:
+
+```python
+pipeline = Embed(Chunk())
+collection = client.create_collection("my_collection", pipeline=pipeline)
+# Add initial data
+collection.add(data=["large_document_1", "large_document_2"], ids=[0, 1])
+
+# Now we just want to add a chunk
+collection.add(data=["custom_chunk"], ids=[2], pipeline=Embed())
+# Or you can pass a pipeline to data
+collection.add(data=Embed({"data": "custom_chunk_2", ids: [3]}))
+```
+Note that you could also do the opposite, where you do not want to chunk the data initially, but later want to chunk the data. For example:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep())
+# Add initial data
+collection.add(data=["small_chunk_1", "small_chunk_2"], ids=[0, 1])
+# Now add large documents with chunking
+collection.add(data=["large_document_1", "large_document_2"], ids=[2, 3], pipeline=DefaultEmbeddingFunctionPipelineStep(Chunk()))
+```
+
+You could also run the pipeline manually and then pass the output to the add step. For example:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep())
+# Add initial data
+collection.add(data=["small_chunk_1", "small_chunk_2"], ids=[0, 1])
+# Now add large documents with chunking but manually run the pipeline
+pipeline = DefaultEmbeddingFunctionPipelineStep(Chunk(["large_document_1", "large_document_2"]))
+data = pipeline.run()
+collection.add(data=data, ids=[2, 3])
+```
+
+#### Server-side registration of pipeline steps
```

Comment:
@HammadB, do you think there's a way to sandbox things? I like the idea of self-contained things, but let's consider unsafe code execution on the server side.

Reply:
So if you are running your own server, its not unsafe, its code you defined.

For hosted chroma, we will have to sandbox the execution.

Reply:
With community-contributed functions this can get messy, if not too complicated perhaps we can offer sandboxed across the board.

Reply:
I think we will have to audit them quite exhaustively for now. I don't think we have to ship with sandboxed in v1